File: src/KernelIntegrator/SPHIntegrationKernel.cc
//---------------------------------Spheral++----------------------------------//
// SPHIntegrationKernel
//----------------------------------------------------------------------------//
#include "SPHIntegrationKernel.hh"

namespace Spheral {

//------------------------------------------------------------------------------
// Constructor
//------------------------------------------------------------------------------
template<typename Dimension>
SPHIntegrationKernel<Dimension>::
SPHIntegrationKernel(const TableKernel<Dimension>& kernel):
  mKernel(kernel) {
}

//------------------------------------------------------------------------------
// Evaluate the functions
//------------------------------------------------------------------------------
template<typename Dimension>
void
SPHIntegrationKernel<Dimension>::
evaluate(const Vector& xp,
         const std::vector<std::pair<int, int>>& indices,
         const FieldList<Dimension, Vector>& position,
         const FieldList<Dimension, SymTensor>& H,
         const FieldList<Dimension, Scalar>& volume,
         const Scalar Hmult,
         std::vector<Scalar>& values,
         std::vector<Vector>& dvalues) const {
  const auto numIndices = indices.size();
  CHECK(values.size() == numIndices
        && dvalues.size() == numIndices);
  for (auto j = 0u; j < numIndices; ++j) {
    auto nodeListj = indices[j].first;
    auto nodej = indices[j].second;
    const auto xj = position(nodeListj, nodej);
    const auto Hj = H(nodeListj, nodej) * Hmult;
    const auto xpj = xp - xj;
    const auto eta = Hj * xpj;
    const auto etaMag = eta.magnitude();
    const auto etaUnit = eta.unitVector();
    const auto Hdet = Hj.Determinant();
    const auto HetaUnit = Hj * etaUnit;
    
    values[j] = mKernel.kernelValue(etaMag, Hdet);
    dvalues[j] = HetaUnit * mKernel.gradValue(etaMag, Hdet);
  }
}

} // end namespace Spheral

File: src/Pybind11Wraps/KernelIntegrator/SPHIntegrationKernel.py
from PYB11Generator import *
from IntegrationKernel import *

@PYB11template("Dimension")
@PYB11holder("std::shared_ptr")
class SPHIntegrationKernel(IntegrationKernel):
    PYB11typedefs = """
    typedef typename %(Dimension)s::Scalar Scalar;
    typedef typename %(Dimension)s::Vector Vector;
    typedef typename %(Dimension)s::SymTensor SymTensor;
"""
    def pyinit(self,
               kernel = "const TableKernel<%(Dimension)s>&"):
        "SPH functions for integration"

    @PYB11virtual
    @PYB11const
    def extent(self,
               Hmult = "const Scalar"):
        "Get extent of kernel"
        return "double"
    
    @PYB11virtual
    @PYB11const
    def evaluate(self,
                 xp = "const Vector&",
                 indices = "const std::vector<std::pair<int, int>>&",
                 position = "const FieldList<%(Dimension)s, Vector>&",
                 H = "const FieldList<%(Dimension)s, SymTensor>&",
                 volume = "const FieldList<%(Dimension)s, Scalar>&",
                 Hmult = "const Scalar",
                 values = "std::vector<Scalar>&",
                 dvalues = "std::vector<Vector>&"):
        "Evaluate some functions at a point"
        return "void"
